# Исследовательский проект

Бершицкий Дмитрий, Пономарев Николай

Данные взяты из https://www.kaggle.com/datasets/nehalbirla/vehicle-dataset-from-cardekho?select=car+details+v4.csv

In [155]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import warnings

warnings.filterwarnings('ignore')

In [156]:
df_source = pd.read_csv('car-selling.csv')
df = df_source.copy()

In [157]:
df_source

Make                                  Model    Price  Year  \
0             Honda                    Amaze 1.2 VX i-VTEC   505000  2017   
1     Maruti Suzuki                        Swift DZire VDI   450000  2014   
2           Hyundai                   i10 Magna 1.2 Kappa2   220000  2011   
3            Toyota                               Glanza G   799000  2019   
4            Toyota        Innova 2.4 VX 7 STR [2016-2020]  1950000  2018   
...             ...                                    ...      ...   ...   
2054       Mahindra                  XUV500 W8 [2015-2017]   850000  2016   
2055        Hyundai                           Eon D-Lite +   275000  2014   
2056           Ford            Figo Duratec Petrol ZXI 1.2   240000  2013   
2057            BMW  5-Series 520d Luxury Line [2017-2019]  4290000  2018   
2058       Mahindra      Bolero Power Plus ZLX [2016-2019]   670000  2017   

      Kilometer Fuel Type Transmission    Location   Color   Owner  \
0         87150    Petrol       Manual        Pune    Grey   First   
1         75000    Diesel       Manual    Ludhiana   White  Second   
2         67000    Petrol       Manual     Lucknow  Maroon   First   
3         37500    Petrol       Manual   Mangalore     Red   First   
4         69000    Diesel       Manual      Mumbai    Grey   First   
...         ...       ...          ...         ...     ...     ...   
2054      90300    Diesel       Manual       Surat   White   First   
2055      83000    Petrol       Manual   Ahmedabad   White  Second   
2056      73000    Petrol       Manual       Thane  Silver   First   
2057      60474    Diesel    Automatic  Coimbatore   White   First   
2058      72000    Diesel       Manual    Guwahati   White   First   

     Seller Type   Engine           Max Power              Max Torque  \
0      Corporate  1198 cc   87 bhp @ 6000 rpm       109 Nm @ 4500 rpm   
1     Individual  1248 cc   74 bhp @ 4000 rpm       190 Nm @ 2000 rpm   
2     Individual  1197 cc   79 bhp @ 6000 rpm  112.7619 Nm @ 4000 rpm   
3     Individual  1197 cc   82 bhp @ 6000 rpm       113 Nm @ 4200 rpm   
4     Individual  2393 cc  148 bhp @ 3400 rpm       343 Nm @ 1400 rpm   
...          ...      ...                 ...                     ...   
2054  Individual  2179 cc  138 bhp @ 3750 rpm       330 Nm @ 1600 rpm   
2055  Individual   814 cc   55 bhp @ 5500 rpm        75 Nm @ 4000 rpm   
2056  Individual  1196 cc   70 bhp @ 6250 rpm       102 Nm @ 4000 rpm   
2057  Individual  1995 cc  188 bhp @ 4000 rpm       400 Nm @ 1750 rpm   
2058  Individual  1493 cc   70 bhp @ 3600 rpm       195 Nm @ 1400 rpm   

     Drivetrain  Length   Width  Height  Seating Capacity  Fuel Tank Capacity  
0           FWD  3990.0  1680.0  1505.0               5.0                35.0  
1           FWD  3995.0  1695.0  1555.0               5.0                42.0  
2           FWD  3585.0  1595.0  1550.0               5.0                35.0  
3           FWD  3995.0  1745.0  1510.0               5.0                37.0  
4           RWD  4735.0  1830.0  1795.0               7.0                55.0  
...         ...     ...     ...     ...               ...                 ...  
2054        FWD  4585.0  1890.0  1785.0               7.0                70.0  
2055        FWD  3495.0  1550.0  1500.0               5.0                32.0  
2056        FWD  3795.0  1680.0  1427.0               5.0                45.0  
2057        RWD  4936.0  1868.0  1479.0               5.0                65.0  
2058        RWD  3995.0  1745.0  1880.0               7.0                 NaN  

[2059 rows x 20 columns]

In [158]:
# Посмотрим на типы данных и наличие пропусков
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2059 entries, 0 to 2058
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Make                2059 non-null   object 
 1   Model               2059 non-null   object 
 2   Price               2059 non-null   int64  
 3   Year                2059 non-null   int64  
 4   Kilometer           2059 non-null   int64  
 5   Fuel Type           2059 non-null   object 
 6   Transmission        2059 non-null   object 
 7   Location            2059 non-null   object 
 8   Color               2059 non-null   object 
 9   Owner               2059 non-null   object 
 10  Seller Type         2059 non-null   object 
 11  Engine              1979 non-null   object 
 12  Max Power           1979 non-null   object 
 13  Max Torque          1979 non-null   object 
 14  Drivetrain          1923 non-null   object 
 15  Length              1995 non-null   float64
 16  Width 

In [159]:
# Посчитаем количество Null значений для каждой переменной
df.isnull().sum()

Make                    0
Model                   0
Price                   0
Year                    0
Kilometer               0
Fuel Type               0
Transmission            0
Location                0
Color                   0
Owner                   0
Seller Type             0
Engine                 80
Max Power              80
Max Torque             80
Drivetrain            136
Length                 64
Width                  64
Height                 64
Seating Capacity       64
Fuel Tank Capacity    113
dtype: int64

# 1. Описание датасета

Датасет содержит 20 столбцов и 2059 строк. В нем содержатся следующие переменные:

| Переменная          | Описание                          | Тип переменной             | Тип данных | Единицы            | Пропуски |
|---------------------|---------------------------------|----------------------------|------------|--------------------|----------|
| Make               | Марка автомобиля               | Категориальная (номинальная) | string     | -                  | 0        |
| Model              | Модель автомобиля             | Категориальная (номинальная) | string     | -                  | 0        |
| Price              | Цена автомобиля               | Метрическая                 | integer    | INR                | 0        |
| Year               | Год выпуска                   | Метрическая                 | integer    | год                | 0        |
| Kilometer          | Пробег автомобиля            | Метрическая                 | integer    | км                 | 0        |
| Fuel Type         | Тип топлива                   | Категориальная (номинальная) | string     | -                  | 0        |
| Transmission       | Тип трансмиссии              | Категориальная (номинальная) | string     | -                  | 0        |
| Location          | Город продажи                 | Категориальная (номинальная) | string     | -                  | 0        |
| Color             | Цвет автомобиля               | Категориальная (номинальная) | string     | -                  | 0        |
| Owner             | Количество владельцев         | Категориальная (порядковая)  | string     | -                  | 0        |
| Seller Type       | Тип продавца                  | Категориальная (номинальная) | string     | -                  | 0        |
| Engine           | Объем двигателя              | Метрическая                 | string     | см³                | 80       |
| Max Power        | Макс. мощность               | Метрическая                 | string     | л.с. @ об/мин      | 80       |
| Max Torque       | Макс. крутящий момент         | Метрическая                 | string     | Нм @ об/мин        | 80       |
| Drivetrain       | Тип привода                   | Категориальная (номинальная) | string     | -                  | 136      |
| Length           | Длина автомобиля             | Метрическая                 | float      | мм                 | 64       |
| Width            | Ширина автомобиля            | Метрическая                 | float      | мм                 | 64       |
| Height           | Высота автомобиля            | Метрическая                 | float      | мм                 | 64       |
| Seating Capacity | Кол-во мест                  | Метрическая                 | float      | -                  | 64       |
| Fuel Tank Capacity | Объем бака                 | Метрическая                 | float      | литры              | 113      |


# 2. Изменение типов данных

Для начала переведем Engine, Max Power и Max Torque в float.

1) Engine содержит значения в виде '1197 cc', '1498 cc' и т.д., поэтому уберем 'cc' и преобразуем в int.
2) Max Power содержит значения в виде '90 bhp @ 6000 rpm', '103 bhp @ 3600 rpm' и т.д., поэтому уберем ' bhp @ 3000 rpm' и преобразуем в int, так как количество оборотов, при котором достигается максимальная мощность, нас не интересует.
3) Max Torque содержит значения в виде '113 Nm @ 4200rpm', '250 Nm @ 1500-2500rpm' и т.д., поэтому уберем ' Nm @ 4200rpm' и преобразуем в int, так как количество оборотов, при котором достигается максимальный крутящий момент, нас не интересует.

In [160]:
# Преобразуем только те строки, которые не NaN
df['Engine'] = df['Engine'].str.replace(' cc', '').dropna().astype(float)
df['Max Power'] = df['Max Power'].str.split('@').str[0].str.replace('bhp', '').dropna().astype(float)
df['Max Torque'] = df['Max Torque'].str.split('@').str[0].str.replace('Nm', '').dropna().astype(float)

Приведем Owner к числовому виду, где 1 - один владелец, 2 - два владельца и т.д.

In [161]:
df['Owner'].unique()

array(['First', 'Second', 'Third', 'Fourth', 'UnRegistered Car',
       '4 or More'], dtype=object)

In [162]:
df[df['Owner'] == 'UnRegistered Car']

Make                         Model     Price  Year  Kilometer  \
139            Audi       Q5 45 TFSI Premium Plus   5651000  2022          1   
194            Audi    Q3 2.0 TDI quattro Premium   1350000  2012      72000   
222            Audi   A6 2.0 TFSi Technology Pack   7200000  2021      20000   
362           Honda            City ZX CVT Petrol   1700000  2022       8530   
389          Toyota                  Innova 2.0 V    575000  2013      82000   
479   Mercedes-Benz   GLC 220d 4MATIC Progressive   6400000  2022       5500   
756   Mercedes-Benz            C-Class 220 CDI AT   5400000  2022       4600   
815         Renault      Kwid 1.0 RXT [2016-2019]    420000  2017      43000   
839   Mercedes-Benz            A-Class Sedan 200d   4100000  2022       3826   
861            Audi       A4 Premium Plus 40 TFSI   4200000  2021      12321   
1533  Mercedes-Benz                   E-Class 200   7050000  2022       1102   
1624  Mercedes-Benz            A-Class Sedan 200d   4200000  2021       5600   
1746           Audi       A4 Premium Plus 40 TFSI   4151000  2022          1   
1766           MINI          Cooper JCW Hatchback   5200000  2022          0   
1782        Renault  Kwid CLIMBER 1.0 [2017-2019]    375000  2017      45000   
1794  Mercedes-Benz            C-Class 220 CDI AT   5300000  2022      12500   
1803        Porsche                  Cayenne Base  14900000  2021       7534   
1840           Audi         A6 Technology 45 TFSI   5800000  2021       7935   
1912  Mercedes-Benz    S-Class S 350D [2018-2020]  18500000  2021       5000   
1977  Mercedes-Benz   GLC 220d 4MATIC Progressive   8100000  2022       5500   
1980  Maruti Suzuki   Wagon R VXi 1.0 [2019-2019]    420000  2018      50000   

     Fuel Type Transmission     Location   Color             Owner  \
139     Petrol    Automatic        Delhi    Blue  UnRegistered Car   
194     Diesel    Automatic    Hyderabad   White  UnRegistered Car   
222     Petrol    Automatic    Bangalore   White  UnRegistered Car   
362     Petrol    Automatic         Pune   White  UnRegistered Car   
389     Petrol       Manual     Varanasi    Gold  UnRegistered Car   
479     Diesel    Automatic    Hyderabad   White  UnRegistered Car   
756     Diesel    Automatic    Bangalore   White  UnRegistered Car   
815     Petrol       Manual    Hyderabad    Grey  UnRegistered Car   
839     Diesel    Automatic    Ahmedabad   White  UnRegistered Car   
861     Petrol    Automatic        Delhi   Black  UnRegistered Car   
1533    Petrol       Manual    Bangalore   Black  UnRegistered Car   
1624    Diesel    Automatic        Delhi    Grey  UnRegistered Car   
1746    Petrol    Automatic        Delhi   Black  UnRegistered Car   
1766    Petrol    Automatic    Ahmedabad  Yellow  UnRegistered Car   
1782    Petrol       Manual  Bhubaneswar    Grey  UnRegistered Car   
1794    Diesel    Automatic    Hyderabad   White  UnRegistered Car   
1803    Petrol    Automatic       Mumbai    Blue  UnRegistered Car   
1840    Petrol    Automatic        Surat   Brown  UnRegistered Car   
1912    Diesel    Automatic         Pune   White  UnRegistered Car   
1977    Diesel    Automatic    Hyderabad   White  UnRegistered Car   
1980    Petrol       Manual  Bhubaneswar   White  UnRegistered Car   

     Seller Type  Engine  Max Power  Max Torque Drivetrain  Length   Width  \
139   Individual  1984.0      248.0       370.0        AWD  4663.0  1898.0   
194   Individual  1968.0      174.0       380.0        AWD  4385.0  1831.0   
222   Individual  1984.0      180.0       320.0        FWD  4915.0  1874.0   
362   Individual  1498.0      119.0       145.0        FWD  4549.0  1748.0   
389   Individual  1998.0      136.0       182.0        NaN  4555.0  1770.0   
479   Individual  1950.0      192.0       400.0        AWD  4658.0  1890.0   
756   Individual  2148.0      170.0       410.0        RWD  4596.0  1770.0   
815   Individual   999.0       67.0        91.0        FWD  3679.0  1579.0   
839   Indiv

Заметим, что есть небольшое количество машин без регистрации, причем они в основном достаточно новые. Поэтому для таких авто будем считать, что у них был 1 владелец.

In [163]:
# Заменим значения в столбце 'Owner' на числовые
df['Owner'] = df['Owner'].replace({
    'First': 1,
    'Second': 2,
    'Third': 3,
    'Fourth': 4,
    'UnRegistered Car': 1,
    '4 or More': 5
})

Посмотрим на изменения в df:

In [164]:
df

Make                                  Model    Price  Year  \
0             Honda                    Amaze 1.2 VX i-VTEC   505000  2017   
1     Maruti Suzuki                        Swift DZire VDI   450000  2014   
2           Hyundai                   i10 Magna 1.2 Kappa2   220000  2011   
3            Toyota                               Glanza G   799000  2019   
4            Toyota        Innova 2.4 VX 7 STR [2016-2020]  1950000  2018   
...             ...                                    ...      ...   ...   
2054       Mahindra                  XUV500 W8 [2015-2017]   850000  2016   
2055        Hyundai                           Eon D-Lite +   275000  2014   
2056           Ford            Figo Duratec Petrol ZXI 1.2   240000  2013   
2057            BMW  5-Series 520d Luxury Line [2017-2019]  4290000  2018   
2058       Mahindra      Bolero Power Plus ZLX [2016-2019]   670000  2017   

      Kilometer Fuel Type Transmission    Location   Color  Owner Seller Type  \
0         87150    Petrol       Manual        Pune    Grey      1   Corporate   
1         75000    Diesel       Manual    Ludhiana   White      2  Individual   
2         67000    Petrol       Manual     Lucknow  Maroon      1  Individual   
3         37500    Petrol       Manual   Mangalore     Red      1  Individual   
4         69000    Diesel       Manual      Mumbai    Grey      1  Individual   
...         ...       ...          ...         ...     ...    ...         ...   
2054      90300    Diesel       Manual       Surat   White      1  Individual   
2055      83000    Petrol       Manual   Ahmedabad   White      2  Individual   
2056      73000    Petrol       Manual       Thane  Silver      1  Individual   
2057      60474    Diesel    Automatic  Coimbatore   White      1  Individual   
2058      72000    Diesel       Manual    Guwahati   White      1  Individual   

      Engine  Max Power  Max Torque Drivetrain  Length   Width  Height  \
0     1198.0       87.0    109.0000        FWD  3990.0  1680.0  1505.0   
1     1248.0       74.0    190.0000        FWD  3995.0  1695.0  1555.0   
2     1197.0       79.0    112.7619        FWD  3585.0  1595.0  1550.0   
3     1197.0       82.0    113.0000        FWD  3995.0  1745.0  1510.0   
4     2393.0      148.0    343.0000        RWD  4735.0  1830.0  1795.0   
...      ...        ...         ...        ...     ...     ...     ...   
2054  2179.0      138.0    330.0000        FWD  4585.0  1890.0  1785.0   
2055   814.0       55.0     75.0000        FWD  3495.0  1550.0  1500.0   
2056  1196.0       70.0    102.0000        FWD  3795.0  1680.0  1427.0   
2057  1995.0      188.0    400.0000        RWD  4936.0  1868.0  1479.0   
2058  1493.0       70.0    195.0000        RWD  3995.0  1745.0  1880.0   

      Seating Capacity  Fuel Tank Capacity  
0                  5.0                35.0  
1                  5.0                42.0  
2                  5.0                35.0  
3                  5.0                37.0  
4                  7.0                55.0  
...                ...                 ...  
2054               7.0                70.0  
2055               5.0                32.0  
2056               5.0                45.0  
2057               5.0                65.0  
2058               7.0                 NaN  

[2059 rows x 20 columns]

# 3. Обработка пропусков

Как мы заметили, пропуски есть у переменных Engine, Max Power, Max Torque, Drivetrain, Length, Width, Height, Seating Capacity, Fuel Tank Capacity.

Заполним все пропуски медианой (в случае числовых переменных) или модой (в случае категориальной Drivetrain).

In [165]:
df['Engine'] = df['Engine'].fillna(df['Engine'].median())
df['Max Power'] = df['Max Power'].fillna(df['Max Power'].median())
df['Max Torque'] = df['Max Torque'].fillna(df['Max Torque'].median())
df['Drivetrain'] = df['Drivetrain'].fillna(df['Drivetrain'].mode()[0])
df['Length'] = df['Length'].fillna(df['Length'].median())
df['Width'] = df['Width'].fillna(df['Width'].median())
df['Height'] = df['Height'].fillna(df['Height'].median())
df['Seating Capacity'] = df['Seating Capacity'].fillna(df['Seating Capacity'].median())
df['Fuel Tank Capacity'] = df['Fuel Tank Capacity'].fillna(df['Fuel Tank Capacity'].median())

In [166]:
df.isnull().sum()

Make                  0
Model                 0
Price                 0
Year                  0
Kilometer             0
Fuel Type             0
Transmission          0
Location              0
Color                 0
Owner                 0
Seller Type           0
Engine                0
Max Power             0
Max Torque            0
Drivetrain            0
Length                0
Width                 0
Height                0
Seating Capacity      0
Fuel Tank Capacity    0
dtype: int64

Больше пропусков нет!

# 4. Новый столбец на основе предыдущих

Рассчитаем для каждой машины среднегодовой пробег. Эта информация может быть полезной, чтобы понять, насколько машина была эксплуатирована.

Автор не указал, в каком году был сделан датасет. Найдем год выпуска самой новой машины и будем считать, что данные актуальны на этот год.

In [167]:
max_year = df['Year'].max()
df['Kilometer Per Year'] = df['Kilometer'] / (max_year + 1 - df['Year'])

# 5. Категориальная переменная из метрической

В датасете есть некоторое количество машин с маленьким пробегом. Вероятно, это машины, которые были куплены, но не использовались или были куплены на перепродажу. Будем считать, что:

- Машина с пробегом менее 1000 км - новая
- Машина с пробегом от 1000 до 10000 км - почти новая
- Машина с пробегом от 10000 - б/у

In [168]:
df['Kilometer Category'] = pd.cut(df['Kilometer'], bins=[-1, 1000, 10000, df['Kilometer'].max()], labels=['New', 'Almost New', 'Used'])

In [169]:
df

Make                                  Model    Price  Year  \
0             Honda                    Amaze 1.2 VX i-VTEC   505000  2017   
1     Maruti Suzuki                        Swift DZire VDI   450000  2014   
2           Hyundai                   i10 Magna 1.2 Kappa2   220000  2011   
3            Toyota                               Glanza G   799000  2019   
4            Toyota        Innova 2.4 VX 7 STR [2016-2020]  1950000  2018   
...             ...                                    ...      ...   ...   
2054       Mahindra                  XUV500 W8 [2015-2017]   850000  2016   
2055        Hyundai                           Eon D-Lite +   275000  2014   
2056           Ford            Figo Duratec Petrol ZXI 1.2   240000  2013   
2057            BMW  5-Series 520d Luxury Line [2017-2019]  4290000  2018   
2058       Mahindra      Bolero Power Plus ZLX [2016-2019]   670000  2017   

      Kilometer Fuel Type Transmission    Location   Color  Owner  ...  \
0         87150    Petrol       Manual        Pune    Grey      1  ...   
1         75000    Diesel       Manual    Ludhiana   White      2  ...   
2         67000    Petrol       Manual     Lucknow  Maroon      1  ...   
3         37500    Petrol       Manual   Mangalore     Red      1  ...   
4         69000    Diesel       Manual      Mumbai    Grey      1  ...   
...         ...       ...          ...         ...     ...    ...  ...   
2054      90300    Diesel       Manual       Surat   White      1  ...   
2055      83000    Petrol       Manual   Ahmedabad   White      2  ...   
2056      73000    Petrol       Manual       Thane  Silver      1  ...   
2057      60474    Diesel    Automatic  Coimbatore   White      1  ...   
2058      72000    Diesel       Manual    Guwahati   White      1  ...   

     Max Power  Max Torque  Drivetrain  Length   Width  Height  \
0         87.0    109.0000         FWD  3990.0  1680.0  1505.0   
1         74.0    190.0000         FWD  3995.0  1695.0  1555.0   
2         79.0    112.7619         FWD  3585.0  1595.0  1550.0   
3         82.0    113.0000         FWD  3995.0  1745.0  1510.0   
4        148.0    343.0000         RWD  4735.0  1830.0  1795.0   
...        ...         ...         ...     ...     ...     ...   
2054     138.0    330.0000         FWD  4585.0  1890.0  1785.0   
2055      55.0     75.0000         FWD  3495.0  1550.0  1500.0   
2056      70.0    102.0000         FWD  3795.0  1680.0  1427.0   
2057     188.0    400.0000         RWD  4936.0  1868.0  1479.0   
2058      70.0    195.0000         RWD  3995.0  1745.0  1880.0   

      Seating Capacity  Fuel Tank Capacity  Kilometer Per Year  \
0                  5.0                35.0        14525.000000   
1                  5.0                42.0         8333.333333   
2                  5.0                35.0         5583.333333   
3                  5.0                37.0         9375.000000   
4                  7.0                55.0        13800.000000   
...                ...                 ...                 ...   
2054               7.0                70.0        12900.000000   
2055               5.0                32.0         9222.222222   
2056               5.0                45.0         7300.000000   
2057               5.0                65.0        12094.800000   
2058               7.0                50.0        12000.000000   

      Kilometer Category  
0                   Used  
1                   Used  
2                   Used  
3                   Used  
4                   Used  
...                  ...  
2054                Used  
2055                Used  
2056                Used  
2057                Used  
2058                Used  

[2059 rows x 22 columns]